### PR TITLE
Issue #3961: DetailASTTest: 'checkTree' failing on deep AST tree

### PR DESCRIPTION
**Pull request by Luolc**
#3961 

The fix will follow the steps below:

- [x] Create an input file and let the CI fail with a stackoverflow error.
- [x] Rewrite the test with a non-recursive way and let the CI pass. To prove the fix will work.
- [x] Create diff report.

~~Currently I am doing step one. I have duplicated the input file which makes CI failed in #3700. I am waiting for the CI result now.~~

The stackoverflow is reproduced successfully. https://travis-ci.org/checkstyle/checkstyle/jobs/209359675#L2123

CI passed after rewriting the test now.

Diff report:
http://www.luolc.com/checkstyle-diff-report/issue3961/

@rnveach please have a review 😄

### CODE REVIEW

File: src/test/java/com/puppycrawl/tools/checkstyle/api/DetailASTTest.java [hunk 1]
```diff
@@ -143,15 +143,30 @@ private static void checkFile(String filename) throws Exception {
...
 
-    private static void checkTree(final DetailAST node,
-                           final DetailAST parent,
-                           final DetailAST prev,
-                           final String filename,
-                           final DetailAST root) {
+    private static void checkTree(final String filename, final DetailAST root) {
+        DetailAST curNode = root;
+        while (curNode != null) {
+            checkNode(curNode, curNode.getParent(), curNode.getPreviousSibling(), filename, root);
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/DetailASTTest.java]: Information sent to the method can't be using the `DetailAST` methods.
The whole point of this test is to verify those exact methods are functioning properly.

Example: Parameter `prev` is set to `curNode.getPreviousSibling()` and inside the method is verified by `assertEquals(badPrevMsg, prev, node.getPreviousSibling());`.
All it is doing now is testing the method against itself. If you change `getPreviousSibling()` to return an invalid value, test would still pass.

Original test kept track of the parent and previous itself. This is why.

Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/DetailASTTest.java]: This test also assumes `getFirstChild` and `getNextSibling` are working correct. We can assume `getParent` and `getPreviousSibling` are valid after this `check` call.

Comment (Luolc) [inline src/test/java/com/puppycrawl/tools/checkstyle/api/DetailASTTest.java]: Done. I misunderstood the purpose of that test before. Sorry for my careless 😅

**Comment by romani** [comment]
uhh, and all that only to keep UT code correct.

can we generate such file in runtime in temp folder ?

**Comment by MEZk** [comment]
@romani 
>  and all that only to keep UT code correct.

Agree. Maybe it is better to skip that one file or just generate it in runtime.

**Comment by romani** [comment]
> On the other hand, it can be placed at resources-noncompilable folder with an explanation why it is there.

it will keep the same size in repo, it very undesirable. Fact that it is NOT a Input for main code lowers its value.

**Comment by Luolc** [comment]
Sorry for the late reply. I went climbing at the last weekend. 😅 

> I am now having second thoughts about keeping the bad input file once I saw it is almost 4 megs in size.

@rnveach I didn't simplify the input file before. The original one is just a duplicate of https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/InputAllEscapedUnicodeCharacters.java. It is 12kb after the optimization, and can also be used to reproduce the stackoverflow error.

@romani @MEZk 
Since it has a much less size now, is it acceptable? If not, I will try to generate it in run time. But I am not very sure how to do that currently, maybe I need some help later.

BTW, I don't know why `wercker/build` failed. My other PR  #3993 failed on it as well and it passed when I had a second try. But this one keeps failing. I could only see a 401 when clicking 'Details' so I am not able to know what's happened.

**Comment by rnveach** [comment]
I am fine with 12kb.

**Comment by romani** [comment]
Changes are ok, but we need to pass wercker.

Wercker logs: https://gist.github.com/romani/306d36ba1a6590658db911906b83ac66
Wercker commands: https://github.com/checkstyle/checkstyle/blob/master/wercker.yml#L172
please run wercker commands on local to reproduce the problem, share the ouput.

**Comment by Luolc** [comment]
This time is `codecov/project` failing... so weird.
Wercker commands on my local ran successfully. I just push again and it passed... Don't know why.
The details of `codecov/project` says that `This issue occurs when the base commit (the parent commit of the first commit on the pull request) did not upload coverage and/or failed CI tests.` It seems that it is a problem of the current head of master. Need help.

**Comment by MEZk** [comment]
@romani 
As all CI validations passed and we all agree with the changes I think we are ready to merge PR.

**Pull request merged**